### ISSUE TITLE
Transcrypt as javascript framework

### ISSUE THREAD

**Issue by alexgustafson** [QQuick/Transcrypt#481]
Hi,

I am using transcrypt for all my frontend stuff and I am really loving it. Thank you for all the great work that went into this project.

I use Transcrypt in conjunction with Angular. Angular helps to wire up the dom with logic in a view/controller fashion with nice two way binding features. This reduces the amount of code I have to write and keeps complexity under control. I'll gladly write up a little description on how I am wrapping Angular in Transcrypt, works nicely with Classes and Mixins.

However, I do have a problem with size of the js files I am loading. A simple Angular Controller or Service written in Javascript might be as small as 800 bytes. The same logic transpiled from python will be 80KB. I could compile the whole app as one big python file using imports, but then I cannot selectively load the controllers and services required on a specific page. I am using Django in the backend and controllers and services are loaded from a Django app's static directory as needed. One one page I might have 4 or 5 controllers with several services or providers, other pages have different combination of modules.

..so, my question. Is it possible to load Transcrypt once as a generic framework on a page, and transpile python files into smaller javascript files that reference the framework?

Thanks again.

**Comment by JdeH** [comment]
You've hit a very real problem!!

- We started out from the "single app per page" paradigm, which was suitable for a number of cases.
- We then were confronted with situations where it was really mandatory to load multiple Transcrypt apps in one page. We simply solved that by insulating their namespaces, which avoided duplicate definitions (but in the process caused bloat by in fact duplicating all library code.

That's were we stand now. Sometimes we use little tricks. E.g. the turtle graphics demo page:

http://www.transcrypt.org/live/turtle_site/turtle_site.html

dynamically loads the code the user made and only compiles that, "inserting" it in the precompiled runtime. This, however is indeed a (rather dirty) trick to limit upload size and compilation time. It shows that what you require is possible IN PRINCIPLE, but can hardly be called a clean solution.

So we have identified the important practiccal need underlying your question, but are still looking for a good solution. It isn't rocket science but rather something that requires some very careful reorganization of Transcrypt not to break anything. But it would greatly enhance its usability to create components.

In short: Thanks for bringing it to the top of our minds again. It will not be solved overnight, but **it's agreed that Transcrypt should be able to do this**! I will dig up and dust off my thinking process on this one. For now unfortunately there's no quick, tidy fix.

Jacques

**Comment by alexgustafson** [comment]
This is great news. Thanks again Jacques for all the hard work that went into this project, it's really such a pleasure to code my frontend logic in Python.

**Comment by BLeAm** [comment]
I've been hitting this limitation with my project as well. While it's useful and benefit many use cases to compile/transpile down to a single monolithic file which included runtime|standard-library per app, in the other hand, it affected the file's size and make redundant codes, the runtime, for instance, inevitable if multiple Transcrypt apps loaded in a page.

It might be a good idea to have a way (pre)compiling the runtime and then compile user's codes as the incremental, if I'm not wrong KotlinJS uses this methodology as well. By this approach, it will widen many use cases and effect dynamic Transcrypt code injection possible.

Moreover, I'd also like to thank you, Jacques, as the founder of Transcrypt, and all contributors, for this very interesting project and the great attitude of you guys.

**Comment by JdeH** [comment]
Precompiling everything except a module written by the user is how the Turtle module already works. Indeed the user code is injected dynamically.

The thing is we'll have to find a concise, future-proof, well-structured way of doing this. The danger of taking a design decision here that unnecessary restricts us in the future is quite large.

So while I agree that we should build this in somehow, I feel the need to take the time needed and  carefully consider how exactly. The pragma's that made Transcrypt lean and fast, were just an idea that came up at some point. Aliases were another idea that enabled us to stay clear of name conflicts while piggybacking our datastructures upon JS ones. Making methods properties by default rather than attributes, thus enabling bound pointer assignment, was another important decision.

An idea of similar quality is needed to have dynamically "linkable" modules in TS. But I am convinced we'll find out how to do that in a solid way. For now all I can tell is that it's on the radar quite large.

**Comment by BLeAm** [comment]
Noted, appreciated, and thank you for sharing. :)

**Comment by JdeH** [comment]
Units / components facilitie added, available from v3.6.101.
Thanks for getting this on the table again!
Feel free to reopen.

**Comment by BLeAm** [comment]
Thanks for updating and your very active development, it's very impressive. One quick question, is there a way to load additional component(s) after the initial load(animals_loader from the document page)? I'm testing the component/unit's concept and looking for some ways to inject components dynamically.

**Comment by JdeH** [comment]
Currently there isn’t. Only a hackish way involving eval, like in the turtle app in the gallery. Eval is too slow for regular use since it makes optimization almost impossible.

**Comment by BLeAm** [comment]
Okay, thank you for your information.
